docs: Send Data Alloy page + Interactive Examples

Jayclifford345 opened the pull request:
This PR updates the [send-data section](https://grafana.com/docs/loki/latest/send-data/) of the documentation with an Alloy introduction and examples directory. Each of the examples are written as interactive sandboxes using the documentation found here:
https://github.com/grafana/killercoda/blob/staging/docs/transformer.md

Note this PR also includes a new writers shortcut
```
{{< docs/ignore >}}
{{/< docs/ignore >}}
```

Here are the Interactive sandboxes auto-generated from the docs:
- https://killercoda.com/grafana-dev-testing/course/loki/alloy-otel-logs
- https://killercoda.com/grafana-dev-testing/course/loki/alloy-kafka-logs


**Checklist**
- [x] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [x] Documentation added
- [x] Documentation contains interactive example
- [ ] Tests updated
- [x] Title matches the required conventional commits format, see [here](https://www.conventionalcommits.org/en/v1.0.0/)
  - **Note** that Promtail is considered to be feature complete, and future development for logs collection will be in [Grafana Alloy](https://github.com/grafana/alloy). As such, `feat` PRs are unlikely to be accepted unless a case can be made for the feature actually being a bug fix to existing behavior.
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [ ] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)
- [ ] If the change is deprecating or removing a configuration option, update the `deprecated-config.yaml` and `deleted-config.yaml` files respectively in the `tools/deprecated-config-checker` directory. [Example PR](https://github.com/grafana/loki/pull/10840/commits/0d4416a4b03739583349934b96f272fb4f685d15)

moxious commented on docs/sources/send-data/alloy/examples/alloy-kafka-logs.md:
```diff
@@ -0,0 +1,373 @@
+
+### Loki Relabel Rules component
+
+Next, we will configure the Loki relabel rules. The `loki.relabel` component rewrites the label set of each log entry passed to its receiver by applying one or more relabeling rules and forwards the results to the list of receivers in the component’s arguments. In our case we are directly calling the rule from the `loki.source.kafka` component.
```
it would help to have a sentence describing why you would want to do this. At the bottom of this section you point out that we're adding `__meta_kafka_topic`.  You might just make some mention of wanting to filter all logs based on those which came from kafka, and that label being a convenient way to do that (or somesuch)

moxious commented on docs/sources/send-data/alloy/examples/alloy-kafka-logs.md:
```diff
@@ -0,0 +1,373 @@
+
+For more information on the `loki.relabel` configuration, see the [Loki Relabel documentation](https://grafana.com/docs/alloy/latest/reference/components/loki.relabel/).
+
+### Loki Write component 
```
See earlier comment: "Send Logs to Loki"

moxious commented on docs/sources/send-data/alloy/examples/alloy-kafka-logs.md:
```diff
@@ -0,0 +1,373 @@
+
+The new configuration will be loaded this can be verified by checking the Alloy UI: [http://localhost:12345](http://localhost:12345).
+
+## Stuck? Need help?
```
Nice touch.

moxious commented on docs/sources/send-data/alloy/examples/alloy-kafka-logs.md:
```diff
@@ -0,0 +1,373 @@
+- Create a few plants to monitor
+- Enable bug mode to activate the bug service. This will cause services to fail and generate additional logs.
+
+Finally to view the logs in Loki, navigate to the Loki Logs Explore view in Grafana at [http://localhost:3000/a/grafana-lokiexplore-app/explore](http://localhost:3000/a/grafana-lokiexplore-app/explore).
```
good use of explore logs

Jayclifford345 commented on docs/sources/send-data/alloy/examples/alloy-kafka-logs.md:
```diff
@@ -0,0 +1,373 @@
+**Note: Killercoda has an inbuilt Code editor which can be accessed via the `Editor` tab.**
+{{< /docs/ignore >}}
+
+### Loki Kafka Source component
```
Great point!! I will get that changed

Jayclifford345 commented:
@JStickler I think it's all ready for your review! No rush at all on this. Let me know if any of the new syntax doesn't make sense I can clear it up

JStickler commented on docs/sources/send-data/alloy/examples/alloy-otel-logs.md:
```diff
@@ -0,0 +1,278 @@
+In this scenario, we have a microservices application called the Carnivourse Greenhouse. This application consists of the following services:
+
+- **User Service:** Mangages user data and authentication for the application. Such as creating users and logging in.
+- **plant Service:** Manges the creation of new plants and updates other services when a new plant is created.
```
```suggestion
- **Plant Service:** Manages the creation of new plants and updates other services when a new plant is created.
```

JStickler commented on docs/sources/send-data/alloy/examples/alloy-otel-logs.md:
```diff
@@ -0,0 +1,278 @@
+
+In this scenario, we have a microservices application called the Carnivourse Greenhouse. This application consists of the following services:
+
+- **User Service:** Mangages user data and authentication for the application. Such as creating users and logging in.
```
```suggestion
- **User Service:** Manages user data and authentication for the application. Such as creating users and logging in.
```

JStickler commented on docs/sources/send-data/alloy/examples/alloy-otel-logs.md:
```diff
@@ -0,0 +1,278 @@
+{{< docs/ignore >}}
+
+### Back to Docs
+Head back to wear you started from to continue with the Loki documentation: [Loki documentation](https://grafana.com/docs/loki/latest/send-data/alloy)
```
```suggestion
Head back to where you started from to continue with the Loki documentation: [Loki documentation](https://grafana.com/docs/loki/latest/send-data/alloy)
```

JStickler reviewed (commented):
Ran through both tutorials in KillerCoda, but most of my comments are on the Kafka file, even though they apply to both.
Let me know if you have any questions.  Nice job on this!

Jayclifford345 commented:
Hey @JStickler! I added this section to make the language more consistent in opening the configuration file. I have also updated the dev-testing versions of the Killercoda sandboxes to match these changes:
https://killercoda.com/grafana-dev-testing/course/loki/alloy-otel-logs
https://killercoda.com/grafana-dev-testing/course/loki/alloy-kafka-logs

Let me know what you think, many thanks again! Sorry if I missed any of your spelling changes in the other page I think I got them all though

JStickler commented on docs/sources/send-data/alloy/examples/alloy-kafka-logs.md:
```diff
@@ -0,0 +1,407 @@
+
+<!-- INTERACTIVE ignore START -->
+1. Open the `loki-fundamentals` directory in a code editor of your choice.
+1. Locate the `config.alloy` file in the `loki-fundamentals` directory (Top level directory).
```
Try to avoid parentheses.
```suggestion
1. Locate the `config.alloy` file in the top level directory, `loki-fundamentals'.
```

JStickler commented on docs/sources/send-data/alloy/examples/alloy-kafka-logs.md:
```diff
@@ -0,0 +1,407 @@
+1. Click on the `config.alloy` file to open it in the code editor.
+<!-- INTERACTIVE ignore END -->
+
+The below configuration snippets will be added to the `config.alloy` file.  
```
Make this active voice.
```suggestion
You will copy all three of the following configuration snippets into the `config.alloy` file.  
```

JStickler commented on docs/sources/send-data/alloy/examples/alloy-kafka-logs.md:
```diff
@@ -0,0 +1,407 @@
+
+Next, we will configure the Loki relabel rules. The `loki.relabel` component rewrites the label set of each log entry passed to its receiver by applying one or more relabeling rules and forwards the results to the list of receivers in the component’s arguments. In our case we are directly calling the rule from the `loki.source.kafka` component.
+
+Add the following configuration to the `config.alloy` file:
```
```suggestion
Now add the following configuration to the `config.alloy` file:
```

JStickler commented on docs/sources/send-data/alloy/examples/alloy-kafka-logs.md:
```diff
@@ -0,0 +1,407 @@
+
+Lastly, we will configure the Loki write component. `loki.write` receives log entries from other loki components and sends them over the network using the Loki logproto format.
+
+Add the following configuration to the `config.alloy` file:
```
```suggestion
And finally, add the following configuration to the `config.alloy` file:
```

JStickler commented on docs/sources/send-data/alloy/examples/alloy-kafka-logs.md:
```diff
@@ -0,0 +1,407 @@
+
+### Open your Code Editor and Locate the `config.alloy` file
+
+Like before, we generate our next pipeline configuration within the same `config.alloy` file. The below configuration snippets will be added **in addition** to the existing configuration. Essentially, we are configuring two pipelines within the same Alloy configuration file.
```
```suggestion
Like before, we generate our next pipeline configuration within the same `config.alloy` file. You will add the following configuration snippets to the file **in addition** to the existing configuration. Essentially, we are configuring two pipelines within the same Alloy configuration file.
```

JStickler commented on docs/sources/send-data/alloy/examples/alloy-otel-logs.md:
```diff
@@ -0,0 +1,293 @@
+1. Click on the `config.alloy` file to open it in the code editor.
+<!-- INTERACTIVE ignore END -->
+
+The below configuration snippets will be added to the `config.alloy` file.  
```
```suggestion
You will copy all three of the following configuration snippets into the `config.alloy` file.
```

JStickler reviewed (commented):
Almost there!

Jayclifford345 commented:
Thanks @JStickler ! That should be those changes added. I have also regenerated the Sandboxes aswell :).

JStickler commented on docs/sources/send-data/alloy/examples/alloy-kafka-logs.md:
```diff
@@ -0,0 +1,406 @@
+- **Websocket Service:** Manages the websocket connections for the application.
+- **Bug Service:** A service that when enabled, randomly causes services to fail and generate additional logs.
+- **Main App:** The main application that ties all the services together.
+- **Database:** A database that stores user and plant data.
```
Did we lose the bullet point for the Plant service?

Jayclifford345 replied:
good catch I must have removed it accepting changes

JStickler commented on docs/sources/send-data/alloy/examples/alloy-otel-logs.md:
```diff
@@ -0,0 +1,292 @@
+In this scenario, we have a microservices application called the Carnivourse Greenhouse. This application consists of the following services:
+
+- **User Service:** Manages user data and authentication for the application. Such as creating users and logging in.
+- **plant Service:** Manages the creation of new plants and updates other services when a new plant is created.
```
```suggestion
- **Plant Service:** Manages the creation of new plants and updates other services when a new plant is created.
```

JStickler commented on docs/sources/send-data/alloy/examples/alloy-otel-logs.md:
```diff
@@ -0,0 +1,292 @@
+
+### Create batches of logs using a OpenTelemetry Processor
+
+Next, we will configure a OpenTelemetry processor. `otelcol.processor.batch` accepts telemetry data from other otelcol components and places them into batches. Batching improves the compression of data and reduces the number of outgoing network requests required to transmit data. This processor supports both size and time based batching.
```
Should otelcol be formatted here?
```suggestion
Next, we will configure a OpenTelemetry processor. `otelcol.processor.batch` accepts telemetry data from other `otelcol` components and places them into batches. Batching improves the compression of data and reduces the number of outgoing network requests required to transmit data. This processor supports both size and time based batching.
```

JStickler commented on docs/sources/send-data/alloy/examples/alloy-otel-logs.md:
```diff
@@ -0,0 +1,292 @@
+
+### Export logs to Loki using a OpenTelemetry Exporter
+
+Lastly, we will configure the OpenTelemetry exporter. `otelcol.exporter.otlphttp` accepts telemetry data from other otelcol components and writes them over the network using the OTLP HTTP protocol. We will use this exporter to send the logs to Loki's native OTLP endpoint.
```
Should otelcol be formatted here?
```suggestion
Lastly, we will configure the OpenTelemetry exporter. `otelcol.exporter.otlphttp` accepts telemetry data from other `otelcol` components and writes them over the network using the OTLP HTTP protocol. We will use this exporter to send the logs to Loki's native OTLP endpoint.
```

JStickler commented on docs/sources/send-data/alloy/examples/alloy-otel-logs.md:
````diff
@@ -0,0 +1,292 @@
+```
+<!-- INTERACTIVE exec END -->
+
+The new configuration will be loaded this can be verified by checking the Alloy UI: [http://localhost:12345](http://localhost:12345).
````
```suggestion
The new configuration will be loaded. You can verify this by checking the Alloy UI: [http://localhost:12345](http://localhost:12345).
```

Jayclifford345 commented on docs/sources/send-data/alloy/examples/alloy-otel-logs.md:
```diff
@@ -0,0 +1,292 @@
+
+It's a fully configured environment with all the dependencies already installed.
+
+![Interactive](/media/docs/loki/loki-ile.svg)
```
I sadly don't think so. When I try putting the image shortcode into the:
```
{{< admonition type="tip" >}}
```
It just returns a bunch of HTML code

Jayclifford345 commented on docs/sources/send-data/alloy/examples/alloy-kafka-logs.md:
```diff
@@ -0,0 +1,406 @@
+
+It's a fully configured environment with all the dependencies already installed.
+
+![Interactive](/media/docs/loki/loki-ile.svg)
```
Yeah sadly I just getting a massive paragraph of HTML tags rather than the image if I try and use the shortcode inside admonition

JStickler reviewed (approved):
[docs team]  LGTM
